[FEATURE] Provide a helpful error message when endpoint is not found

**Is your feature request related to a problem? Please describe.**
It might be not clear what the problem is when I try to look up for an endpoint that doesn't exist in the schema:

```
>>> schema["/userz"]["POST"]

E   KeyError: '/userz/'
```

And if the endpoint path is long it could be hard to detect a typo

**Describe the solution you'd like**
Provide a hint about what is the closest option (via `difflib` for example)

For example:

```
>>> schema["/userz"]["POST"]

E   KeyError: No such endpoint: `POST /userz/`. Did you mean `POST /users/`?
```

The implementation is to replace `CaseInsensitiveDict` with a custom one, that raises a proper error message

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -12,6 +12,7 @@ Changelog
   in your schema or customize existing (like ``application/json``).
 - The `response_schema_conformance` check now runs on media types that are encoded with JSON. For example, ``application/problem+json``. `#920`_
 - Base URL for GraphQL schemas. It allows you to load the schema from one place but send test requests to another one. `#934`_
+- A helpful error message when an operation is not found during the direct schema access. `#812`_
 
 **Changed**
 
@@ -1654,6 +1655,7 @@ Deprecated
 .. _#824: https://github.com/schemathesis/schemathesis/issues/824
 .. _#816: https://github.com/schemathesis/schemathesis/issues/816
 .. _#814: https://github.com/schemathesis/schemathesis/issues/814
+.. _#812: https://github.com/schemathesis/schemathesis/issues/812
 .. _#795: https://github.com/schemathesis/schemathesis/issues/795
 .. _#793: https://github.com/schemathesis/schemathesis/issues/793
 .. _#788: https://github.com/schemathesis/schemathesis/issues/788

--- a/src/schemathesis/schemas.py
+++ b/src/schemathesis/schemas.py
@@ -7,6 +7,7 @@ Their responsibilities:
 They give only static definitions of endpoints.
 """
 from collections.abc import Mapping
+from difflib import get_close_matches
 from typing import Any, Callable, Dict, Generator, Iterable, Iterator, List, Optional, Sequence, Tuple, Type, Union
 from urllib.parse import urljoin, urlsplit, urlunsplit
 
@@ -23,6 +24,16 @@ from .models import Case, Endpoint
 from .stateful import APIStateMachine, Stateful, StatefulTest
 from .types import Filter, FormData, GenericTest, NotSet
 from .utils import NOT_SET, GenericResponse
+
+
+class MethodsDict(CaseInsensitiveDict):
+    def __getitem__(self, item: Any) -> Any:
+        try:
+            return super().__getitem__(item)
+        except KeyError as exc:
+            available_methods = ", ".join(map(str.upper, self))
+            message = f"Method `{item}` not found. Available methods: {available_methods}"
+            raise KeyError(message) from exc
 
 
 @attr.s()  # pragma: no mutate
@@ -46,8 +57,15 @@ class BaseSchema(Mapping):
     def __iter__(self) -> Iterator[str]:
         return iter(self.endpoints)
 
-    def __getitem__(self, item: str) -> CaseInsensitiveDict:
-        return self.endpoints[item]
+    def __getitem__(self, item: str) -> MethodsDict:
+        try:
+            return self.endpoints[item]
+        except KeyError as exc:
+            matches = get_close_matches(item, list(self.endpoints))
+            message = f"`{item}` not found"
+            if matches:
+                message += f". Did you mean `{matches[0]}`?"
+            raise KeyError(message) from exc
 
     def __len__(self) -> int:
         return len(self.endpoints)
@@ -88,7 +106,7 @@ class BaseSchema(Mapping):
         return self._build_base_url()
 
     @property
-    def endpoints(self) -> Dict[str, CaseInsensitiveDict]:
+    def endpoints(self) -> Dict[str, MethodsDict]:
         if not hasattr(self, "_endpoints"):
             # pylint: disable=attribute-defined-outside-init
             endpoints = self.get_all_endpoints()
@@ -267,9 +285,9 @@ class BaseSchema(Mapping):
         raise NotImplementedError
 
 
-def endpoints_to_dict(endpoints: Generator[Endpoint, None, None]) -> Dict[str, CaseInsensitiveDict]:
-    output: Dict[str, CaseInsensitiveDict] = {}
+def endpoints_to_dict(endpoints: Generator[Endpoint, None, None]) -> Dict[str, MethodsDict]:
+    output: Dict[str, MethodsDict] = {}
     for endpoint in endpoints:
-        output.setdefault(endpoint.path, CaseInsensitiveDict())
+        output.setdefault(endpoint.path, MethodsDict())
         output[endpoint.path][endpoint.method] = endpoint
     return output

--- a/test/test_models.py
+++ b/test/test_models.py
@@ -1,3 +1,4 @@
+import re
 from test.utils import SIMPLE_PATH
 
 import pytest
@@ -267,3 +268,20 @@ def test_from_case(swagger_20, base_url, expected):
     session = requests.Session()
     request = Request.from_case(case, session)
     assert request.uri == "http://127.0.0.1/api/v3/users/test"
+
+
+@pytest.mark.parametrize(
+    "value, message",
+    (
+        ("/userz", "`/userz` not found. Did you mean `/users`?"),
+        ("/what?", "`/what?` not found"),
+    ),
+)
+def test_endpoint_path_suggestion(swagger_20, value, message):
+    with pytest.raises(KeyError, match=re.escape(message)):
+        swagger_20[value]["POST"]
+
+
+def test_method_suggestion(swagger_20):
+    with pytest.raises(KeyError, match="Method `PUT` not found. Available methods: GET"):
+        swagger_20["/users"]["PUT"]

